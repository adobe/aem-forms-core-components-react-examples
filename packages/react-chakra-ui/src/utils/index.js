import React from 'react';
import { FormContext } from '@aemforms/af-react-renderer';
import { createFormInstance } from '@aemforms/af-core';
import { ChakraProvider } from '@chakra-ui/react'
import { IntlProvider } from 'react-intl';
import { render } from '@testing-library/react';

export const DEFAULT_ERROR_MESSAGE = 'There is an error in the field';

export const createForm = (field) => {
  const formJson = {
    items: [field]
  };
  return createFormInstance(formJson);
};

export const Provider = (form, mappings, locale, dictionaries) => props => {
  const c = {
    form,
    mappings,
    modelId: '$form',
    refMap: {}
  };
  const { children } = props;
  return (
    <ChakraProvider>
      <IntlProvider locale={locale} messages={dictionaries}>
        <FormContext.Provider value={c}>{children}</FormContext.Provider>
      </IntlProvider>
    </ChakraProvider>
  );
};

export const renderComponent = function (Component) {
  const test = (field, operation) => {
    const form = createForm(field);
    if (operation) {
      operation(form, form.items[0]);
    }
    const e = form.items[0].getState();
    let component = <Component {...e} />;
    const wrapper = Provider(form);
    const renderResponse = render(component, { wrapper });
    return {
      renderResponse,
      form,
      element: form?.items[0],
      component
    };
  };
  return test;
};
