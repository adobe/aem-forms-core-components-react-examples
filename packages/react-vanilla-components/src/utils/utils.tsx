export const formatBytes = (bytes: number, decimals = 0) => {
    if (!+bytes) {
      return '0 Bytes';
    }
    const k = 1024;
    const dm = decimals < 0 ? 0 : decimals;
    const sizes = ['bytes', 'kb', 'mb', 'gb', 'tb', 'pb'];
    const i = Math.floor(Math.log(bytes) / Math.log(k));

    return `${parseFloat((bytes / Math.pow(k, i)).toFixed(dm))} ${sizes[i]}`;
  };

  export const syncAriaDescribedBy = (id: string, tooltip?: string, description?: string, error?: string) => {
    const descriptions = [];

    if (description) {
        descriptions.push(`${id}__longdescription`);
    }
    if (tooltip) {
        descriptions.push(`${id}__shortdescription`);
    }
    if (error) {
        descriptions.push(`${id}__errormessage`);
    }

    return descriptions.join(' ');
};