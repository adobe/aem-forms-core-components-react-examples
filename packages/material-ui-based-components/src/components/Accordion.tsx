import React, { useContext } from 'react';
import { getRenderer, FormContext } from '@aemforms/af-react-renderer';
import { withRuleEnginePanel } from '../shared/withRuleEngine';
import { PROPS_PANEL } from '../utils/types';

import Accordion from '@mui/material/Accordion';
import AccordionSummary from '@mui/material/AccordionSummary';
import AccordionDetails from '@mui/material/AccordionDetails';
import Typography from '@mui/material/Typography';
import ExpandMoreIcon from '@mui/icons-material/ExpandMore';

const AccordionComponent = function (props: PROPS_PANEL) {
    const [expanded, setExpanded] = React.useState<number>(0);
    // @ts-ignore
    const mappings = useContext(FormContext).mappings;
    const items = props.items || [];

    const handleChange = (index: number) => {
        setExpanded(index);
    };

    const getChild = (child: any, index: number) => {
        const Comp = getRenderer(child, mappings);
        return Comp ? <Comp key={`${child.id}_${index}`} {...child} /> : (null);
    };


    return (
        <div>
            <div>{props?.label?.value ? <div>{props.label.value}</div> : null}</div>
            {
                items.map((child: any, index: number) => (
                    <Accordion expanded={expanded === index} onChange={() => { handleChange(index); }} key={`${child.id}_${index}`}>
                        <AccordionSummary
                            expandIcon={<ExpandMoreIcon />}
                            id={child.id}
                        >
                            <Typography>
                                {child.label.value}
                            </Typography>
                        </AccordionSummary>
                        <AccordionDetails>
                            {getChild(child, index)}
                        </AccordionDetails>
                    </Accordion>
                ))
            }
        </div>
    );
};

export default withRuleEnginePanel(AccordionComponent);