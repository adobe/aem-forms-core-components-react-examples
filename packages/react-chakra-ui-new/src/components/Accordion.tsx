/*************************************************************************
* ADOBE CONFIDENTIAL
* ___________________
*
* Copyright 2023 Adobe
* All Rights Reserved.
*
* NOTICE: All information contained herein is, and remains
* the property of Adobe and its suppliers, if any. The intellectual
* and technical concepts contained herein are proprietary to Adobe
* and its suppliers and are protected by all applicable intellectual
* property laws, including trade secret and copyright laws.
* Dissemination of this information or reproduction of this material
* is strictly forbidden unless prior written permission is obtained
* from Adobe.

* Adobe permits you to use and modify this file solely in accordance with
* the terms of the Adobe license agreement accompanying it.
*************************************************************************/

import React, { useCallback, useContext } from 'react';
import { getRenderer, FormContext } from '@aemforms/af-react-renderer';
import { withRuleEnginePanel, richTextString } from '../shared/withRuleEngine';
import { PROPS_PANEL } from '../utils/type';
import {
    Accordion,
    AccordionItem,
    AccordionButton,
    AccordionPanel,
    AccordionIcon,
    Box
  } from '@chakra-ui/react';

const AccordionComponent = function (props: PROPS_PANEL) {
    const [expanded, setExpanded] = React.useState<number>(0);
    // @ts-ignore
    const mappings = useContext(FormContext).mappings;
    const items = props.visible ? props.items : [];

    const handleChange = useCallback((index: number) => {
        setExpanded(index);
    }, []);

    const getChild = useCallback((child: any, index: number) => {
        const Comp = getRenderer(child, mappings);
        return Comp ? <Comp key={`${child.id}_${index}`} {...child} /> : (null);
    }, []);

    return (
        <div>
            {props?.label?.visible ? <div>{props.label.value}</div> : null}
            {
                <Accordion allowToggle index={expanded} onChange={(index: number) => { handleChange(index); }} isDisabled={!props.enabled}>
                    <AccordionItem>
                      <h2>
                        <AccordionButton>
                          <Box flex='1' textAlign='left'>
                            {props?.label?.richText ? richTextString(props?.label?.value) : props?.label?.value}
                          </Box>
                          <AccordionIcon />
                        </AccordionButton>
                      </h2>
                      <AccordionPanel pb={4}>
                        {
                            items.map((child: any, index: number) => (
                                getChild(child, index)
                            ))
                        }
                      </AccordionPanel>
                    </AccordionItem>
                  </Accordion>
            }
        </div>
    );
};

export default withRuleEnginePanel(AccordionComponent);