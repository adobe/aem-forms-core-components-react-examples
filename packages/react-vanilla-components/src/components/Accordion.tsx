// *******************************************************************************
//  * Copyright 2022 Adobe
//  *
//  * Licensed under the Apache License, Version 2.0 (the “License”);
//  * you may not use this file except in compliance with the License.
//  * You may obtain a copy of the License at
//  *
//  *     http://www.apache.org/licenses/LICENSE-2.0
//  *
//  * Unless required by applicable law or agreed to in writing, software
//  * distributed under the License is distributed on an “AS IS” BASIS,
//  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  * See the License for the specific language governing permissions and
//  * limitations under the License.

//  *  The BEM markup is as per the AEM core form components guidelines.
//  * LINK- https://github.com/adobe/aem-core-forms-components/blob/master/ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/accordion/v1/accordion/accordion.html
//  ******************************************************************************

import React, { useState, useCallback, useContext } from 'react';
import { getRenderer, FormContext } from '@aemforms/af-react-renderer';
import { withRuleEnginePanel } from '../utils/withRuleEngine';
import { PROPS_PANEL } from '../utils/type';

const Accordion = (props: PROPS_PANEL) => {
  // @ts-ignore
  const { mappings } = useContext(FormContext);
  const { items, id, label, visible, enabled } = props;
  const [activePanel, setActivePanel] = useState(0);
  const [shortDescription, setShortDescription] = useState(true);
  const [longDescription, setLongtDescription] = useState(false);
  const { v: visibleItems } =
    items.reduce(({ v }: any, item) => {
      const isVisible = item.visible === true;
      return {
        v: isVisible ? v.concat([item]) : v
      };
    }, { v: [] });

  const handleToggle = (i: any) => {
    setActivePanel(i === activePanel ? null : i);
  };

  const handleClick = (event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
    event.preventDefault();
    setShortDescription(!shortDescription);
    setLongtDescription(!longDescription);
  };

  const getChild = useCallback((child: any, index: number) => {
    const Comp = getRenderer(child, mappings);
    return Comp ? <Comp key={`${child.id}_${index}`} {...child} /> : (null);
  }, []);

  return (
    <div className="cmp-accordion" data-cmp-is="adaptiveFormAccordion" data-cmp-single-expansion="true" data-placeholder-text="false" data-cmp-visible={visible} data-cmp-enabled={enabled}>
      {label?.visible && <label htmlFor={id} className="cmp-accordion__label">{label?.value}</label>}
      {props?.description && <button aria-label='Toggle Button' className="cmp-accordion__questionmark" onClick={handleClick}></button>}
      {shortDescription && props?.tooltip && <div title='Help Text' data-cmp-visible={shortDescription} className='cmp-accordion__shortdescription'>{props?.tooltip}</div>}
      <div aria-live="polite">
        {longDescription && props?.description && <div title='Help Text' data-cmp-visible={longDescription} className="cmp-accordion__longdescription">{props?.description}</div>}
      </div>
      {visibleItems?.map((item: PROPS_PANEL, index: number) => (
        <div key={item.id} className="cmp-accordion__item" data-cmp-hook-accordion="item" id={`${item.id}-widget`}>
          <h3 className="cmp-accordion__header">
            <button aria-controls={`${item.id}`} role='button' id={`${item.id}-button`} data-cmp-hook-accordion="button" type="button" onClick={() => handleToggle(index)} className={activePanel === index ? 'cmp-accordion__button cmp-accordion__button--expanded' : 'cmp-accordion__button'} aria-expanded={activePanel === index ? 'true' : 'false'}>
              <span className="cmp-accordion__title">{item?.label?.value}</span>
              <span className="cmp-accordion__icon"></span>
            </button>
          </h3>
          <div className={activePanel === index ? 'cmp-accordion__panel cmp-accordion__panel--expanded' : 'cmp-accordion__panel cmp-accordion__panel--hidden'} data-cmp-hook-adaptiveformaccordion="panel" id={`${item.id}-panel`} role="region" aria-labelledby={`${item.id}-button`} aria-hidden={activePanel === index ? 'false' : 'true'}>
            {activePanel === index && <div className="panelcontainer responsivegrid">
              <div className="cmp-container" data-cmp-is="adaptiveFormPanel" data-cmp-visible={activePanel === index ? 'true' : 'false'} data-cmp-enabled={activePanel === index ? 'true' : 'false'}>
                <label htmlFor={item.id} className="cmp-container__label">{item?.label?.value}</label>
                <div className="aem-Grid aem-Grid--12 aem-Grid--default--12 ">
                  <div className="cmp-adaptiveform-checkboxgroup aem-GridColumn aem-GridColumn--default--12">
                    {activePanel !== null ? getChild(item, index) : null}
                  </div>
                </div>
              </div>
            </div>}
          </div>
        </div>
      ))}
    </div>
  );
};

export default withRuleEnginePanel(Accordion);