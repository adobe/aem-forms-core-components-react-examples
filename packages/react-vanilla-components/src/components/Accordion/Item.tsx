// *******************************************************************************
//  * Copyright 2023 Adobe
//  *
//  * Licensed under the Apache License, Version 2.0 (the “License”);
//  * you may not use this file except in compliance with the License.
//  * You may obtain a copy of the License at
//  *
//  *     http://www.apache.org/licenses/LICENSE-2.0
//  *
//  * Unless required by applicable law or agreed to in writing, software
//  * distributed under the License is distributed on an “AS IS” BASIS,
//  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  * See the License for the specific language governing permissions and
//  * limitations under the License.
//  ******************************************************************************

import React, { useContext, useCallback } from 'react';
import { FormContext, getRenderer } from '@aemforms/af-react-renderer';
import { AccordionContext } from './Accordion';

const Item = (props: any) => {
  const { onAdd, onRemove, isRepeatable, showAddButton, showDeleteButton } = props;
  // @ts-ignore
  const { mappings } = useContext(FormContext);
  const { onToggle, activePanel } = useContext(AccordionContext);
  const getChild = useCallback((child: any) => {
    const Comp = getRenderer(child, mappings);
    return Comp ? <Comp key={`${child.id}-child`} {...child} /> : (null);
  }, []);

  return (
    <div
      className="cmp-accordion__item"
      data-cmp-hook-accordion="item"
      id={`${props.id}-widget`}
    >
      <h3 className="cmp-accordion__header">
        <button
          aria-controls={`${props.id}`}
          role='button'
          id={`${props.id}-button`}
          data-cmp-hook-accordion="button"
          type="button"
          onClick={() => onToggle(props.id)}
          className={`cmp-accordion__button ${activePanel === props.id ? 'cmp-accordion__button--expanded' : ''}`}
          aria-expanded={activePanel === props.id ? 'true' : 'false'}
        >
          <span className="cmp-accordion__title">{props?.label?.value}</span>
          <span className="cmp-accordion__icon"></span>
        </button>
        {
          isRepeatable ? (
            <div
              className="cmp-accordion__repeatable-buttons"
              data-cmp-hook-adaptiveformaccordion="repeatableButton"
            >
              {
                showAddButton &&
                <button
                  className="cmp-accordion__add-button"
                  type="button"
                  data-cmp-visible="true"
                  onClick={onAdd}
                ></button>
              }
              {
                showDeleteButton &&
                <button
                  className="cmp-accordion__remove-button"
                  type="button"
                  data-cmp-visible="true"
                  onClick={onRemove}
                ></button>
              }
            </div>
          ) : null
        }
      </h3>
      <div
        className={`cmp-accordion__panel cmp-accordion__panel--${activePanel === props.id ? 'expanded' : 'hidden'}`}
        data-cmp-hook-adaptiveformaccordion="panel"
        id={`${props.id}-panel`}
        role="region"
        aria-labelledby={`${props.id}-button`}
        aria-hidden={activePanel === props.id ? 'false' : 'true'}
      >
        {activePanel !== null ? getChild(props) : null}
      </div>
    </div>
  );
};

export default Item;