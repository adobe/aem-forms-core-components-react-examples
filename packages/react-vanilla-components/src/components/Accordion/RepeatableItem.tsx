// *******************************************************************************
//  * Copyright 2023 Adobe
//  *
//  * Licensed under the Apache License, Version 2.0 (the “License”);
//  * you may not use this file except in compliance with the License.
//  * You may obtain a copy of the License at
//  *
//  *     http://www.apache.org/licenses/LICENSE-2.0
//  *
//  * Unless required by applicable law or agreed to in writing, software
//  * distributed under the License is distributed on an “AS IS” BASIS,
//  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  * See the License for the specific language governing permissions and
//  * limitations under the License.
//  ******************************************************************************

import React, { useContext } from 'react';
import { AddItem, RemoveItem } from '@aemforms/af-core';
import { FormContext } from '@aemforms/af-react-renderer';
import { withRuleEnginePanel } from '../../utils/withRuleEngine';
import Item from './Item';
import { PROPS_PANEL } from '../../utils/type';

const RepeatableItem = (props: PROPS_PANEL) => {
  // @ts-ignore
  const { form } = useContext(FormContext);
  const element = form.getElement(props.id);
  const repeatableItem = element.getState().items;

  const addPanel = () => {
    element.dispatch(new AddItem());
  };

  const removePanel = () => {
    element.dispatch(new RemoveItem());
  };

  const showAddButton = (repeatableItem.length !== props.maxItems);

  const showDeleteButton = (repeatableItem.length !== props.minItems);

  return (
    repeatableItem.map((item: any) => (
      <Item key={`${item.id}-repeatable`} {...item} onAdd={addPanel} onRemove={removePanel} isRepeatable showAddButton={showAddButton} showDeleteButton=
        {showDeleteButton} />
    ))
  );
};

export default withRuleEnginePanel(RepeatableItem);