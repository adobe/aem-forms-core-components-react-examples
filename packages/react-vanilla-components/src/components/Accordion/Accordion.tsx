// *******************************************************************************
//  * Copyright 2023 Adobe
//  *
//  * Licensed under the Apache License, Version 2.0 (the “License”);
//  * you may not use this file except in compliance with the License.
//  * You may obtain a copy of the License at
//  *
//  *     http://www.apache.org/licenses/LICENSE-2.0
//  *
//  * Unless required by applicable law or agreed to in writing, software
//  * distributed under the License is distributed on an “AS IS” BASIS,
//  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  * See the License for the specific language governing permissions and
//  * limitations under the License.

//  *  The BEM markup is as per the AEM core form components guidelines.
//  * LINK- https://github.com/adobe/aem-core-forms-components/blob/master/ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/form/accordion/v1/accordion/accordion.html
//  ******************************************************************************

import React, {useState, createContext} from 'react';
import { withRuleEnginePanel }  from '../../utils/withRuleEngine';
import { PROPS_PANEL } from '../../utils/type';
import Item from './Item';
import RepeatableItem from './RepeatableItem';

export const AppContext = createContext('' as any);

const Accordion = (props: PROPS_PANEL) => {
   const {items,id, label, visible, enabled} = props;
   const [shortDescription, setShortDescription] = useState(true);
   const [longDescription, setLongtDescription] = useState(false);

   const {v: visibleItems} =
   items.reduce(({v} : any, item) => {
      const isVisible = item.visible === true;
      return {
         v: isVisible ? v.concat([item]) : v
      };
   }, {v: []}); 

   const firstItem = visibleItems[0] || {};

   // for active first item
   let panelId = firstItem?.type === 'array' ?  firstItem?.items[0]?.id : firstItem?.id;
   const [activePanel, setActivePanel] = useState(panelId) ;

   const handleToggle = (id: string) => {
      setActivePanel(id === activePanel ? null : id);
   };

   const handleClick = (event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
      event.preventDefault();  
      setShortDescription(!shortDescription);
      setLongtDescription(!longDescription);
   };

   return (
      <div className="cmp-accordion" data-cmp-is="adaptiveFormAccordion" data-cmp-single-expansion="true" data-placeholder-text="false" data-cmp-visible={visible} data-cmp-enabled={enabled}>
         {label?.visible && <label htmlFor={id} className="cmp-accordion__label">{label?.value}</label>}
         {props?.description && <button aria-label='Toggle Button' className="cmp-accordion__questionmark" onClick={handleClick}></button>} 
         {shortDescription && props?.tooltip && <div title='Help Text' data-cmp-visible={shortDescription} className='cmp-accordion__shortdescription'>{props?.tooltip}</div>}
          <div aria-live="polite">
             {longDescription && props?.description  && <div title='Help Text' data-cmp-visible={longDescription} className="cmp-accordion__longdescription">{props?.description}</div>}
         </div>
          {visibleItems?.map((item: PROPS_PANEL) => {
            return (item?.type === 'array') ?
            //@ts-ignore
            <AppContext.Provider value={{onToggle:handleToggle, activePanel:activePanel}}>
            <RepeatableItem {...item}/> 
            </AppContext.Provider> :  <AppContext.Provider value={{onToggle:handleToggle, activePanel:activePanel}}><Item {...item} /></AppContext.Provider>;
            })}
      </div>
   ); 
};

export default withRuleEnginePanel(Accordion);