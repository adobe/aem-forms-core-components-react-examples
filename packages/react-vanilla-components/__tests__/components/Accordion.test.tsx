/*
 * Copyright 2023 Adobe, Inc.
 *
 * Your access and use of this software is governed by the Adobe Customer Feedback Program Terms and Conditions or other Beta License Agreement signed by your employer and Adobe, Inc.. This software is NOT open source and may not be used without one of the foregoing licenses. Even with a foregoing license, your access and use of this file is limited to the earlier of (a) 180 days, (b) general availability of the product(s) which utilize this software (i.e. AEM Forms), (c) January 1, 2023, (d) Adobe providing notice to you that you may no longer use the software or that your beta trial has otherwise ended.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL ADOBE NOR ITS THIRD PARTY PROVIDERS AND PARTNERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import Accordion from "../../src/components/Accordion/Accordion"
import { renderComponent } from "../utils";
import userEvent from "@testing-library/user-event";

const AccordionWithData = {
    id: "accordion",
    ":type": "core/fd/components/form/accordion/v1/accordion",
    visible: true,
    label: { value: "Accordion" },
    index: 0,
    items: [
        {
          id: "panelcontainer-541adbc656",
          fieldType: "panel",
          name: "item_1",
          visible: true,
          enabled: true,
          items: [
            {
              id: "checkboxgroup-5f5468260f",
              fieldType: "checkbox-group",
              name: "checkboxgroup1675921070268",
              visible: true,
              type: "number[]",
              required: false,
              enumNames: [
                "Male",
                "Female"
              ],
              label: {
                value: "Check Box Group",
                visible: true,
              },
              enum: [
                0,
                1
              ],
            },
          ],
          label: {
            value: "Items",
            visible: true,

          },
        },
        {
          id: "panelcontainer-e48ec263e1",
          fieldType: "panel",
          name: "item_3",
          visible: true,
          enabled: true,
          items: [
            {
              id: "numberinput-d5f36deec8",
              fieldType: "number-input",
              name: "item_16824325701471682432584410",
              visible: true,
              type: "string",
              label: {
                value: "Number Input",
                visible: true,
              },
            }
          ],
          label: {
            value: "Number",
            visible: true,
          }
        }
      ],
  };

  const AccordionwithRepeatData = { "fieldType": "panel",
  ":type": "core/fd/components/form/accordion/v1/accordion",
  "name": "guests",
  "items": [
      {
          "fieldType": "panel",
          "name": "guestDetails",
          "repeatable": true,
          "minOccur": 1,
          "maxOccur": 3,
          "label": {
              "value": "Guest Details",
              "visible": false
          },
          "items": [
            {
              "fieldType": "panel",
              "items": [
                {
                  "fieldType": "text-input",
                  "name": "guestName",
                  "label": {
                      "value": "Full Name"
                  },
                  "required": true,
                  "constraintMessages": {
                      "required": "Required field"
                  }
              },
              {
                  "fieldType": "text-input",
                  "name": "guestAadhar",
                  "label": {
                      "value": "Enter your <b>Aadhar Number</b>",
                      "richText": true
                  },
                  "format": "XXXX XXXX XXXX",
                  "maxLength": 14,
                  "description": "Your format is XXXX XXXX XXXX",
                  "placeholder": "XXXX XXXX XXXX",
                  "required": true,
                  "constraintMessages": {
                      "required": "Required field"
                  }
              }
              ]
            }
          ]
      }]}

  test("First tab should open by default and on clicking it should get collapse", () => {
     const helper = renderComponent(Accordion);
     const { renderResponse } = helper(AccordionWithData);
     const button = renderResponse.container.getElementsByClassName('cmp-accordion__button');
     expect(button).toHaveLength(2);
     expect(button[0].className.includes("cmp-accordion__button cmp-accordion__button--expanded"))
     expect(button[1].className.includes("cmp-accordion__button"));
     userEvent.click(button[0])
     expect(button[0].className.includes("cmp-accordion__button"))
  });

  test('Label of accordion should be visible', () => {
    const helper = renderComponent(Accordion);
    const { renderResponse } = helper(AccordionWithData);
    expect(renderResponse.queryByText(AccordionWithData.label.value)).not.toBeNull();
  });

  test('Individual label should be visible', () => {
    const helper = renderComponent(Accordion);
    const { renderResponse } = helper(AccordionWithData);
    const title = renderResponse.container.getElementsByClassName('cmp-accordion__title')
    expect(title).toHaveLength(2);
    expect(title).not.toBeNull();
  });

  test('In case of both tooltip and description, tooltip should be visible and onclick of toggle button, description should be visible', async () => {
    const f = {
      ...AccordionWithData,
      tooltip: 'Short Description',
      description: 'Mandatory',
      properties: {
        "afs:layout": {
            tooltipVisible: true
        },
      }
    };
    const helper = renderComponent(Accordion);
     const { renderResponse } = await helper(f);
     expect(renderResponse.getByText('Short Description')).not.toBeNull();
     const button = renderResponse.container.getElementsByClassName('cmp-accordion__questionmark');
     userEvent.click(button[0]);
     expect(renderResponse.getByText('Mandatory')).not.toBeNull();
  });

  test('html in the label should be handled for non rich text', async () => {
    const helper = renderComponent(Accordion);
    const f =  {
      ...AccordionWithData,
      label: {
          value: '<p>title inside p tags</p>',
          richText: true,
          visible: true
      }
  }
   let {renderResponse} = await helper(f);
   expect(renderResponse.container.innerHTML).toContain( '<p>title inside p tags</p>');  
  });   

  test('On click of add button, panel should be added and by reaching to max item limit, add button should get disappear', () => {
    const helper = renderComponent(Accordion);
    const { renderResponse } = helper(AccordionwithRepeatData);
    const button = renderResponse.container.getElementsByClassName("cmp-accordion__add-button");
    expect(button.length).toBe(1);
    const panel = renderResponse.container.getElementsByClassName("cmp-accordion__item")
    userEvent.click(button[0]);
    expect(button.length).toBe(2)
    expect(panel.length).toBe(2)
    userEvent.click(button[1])
    expect(panel.length).toBe(3);
    expect(button.length).toBe(0)
  });

  test('Initially remove button should not be visible in the panel and when panel is added, it must appear', () => {
    const helper = renderComponent(Accordion);
    const { renderResponse } = helper(AccordionwithRepeatData);
    const AddButton = renderResponse.container.getElementsByClassName("cmp-accordion__add-button");
    expect(AddButton.length).toBe(1);
    const RemoveButton = renderResponse.container.getElementsByClassName("cmp-accordion__remove-button");
    expect(RemoveButton.length).toBe(0);
    const panel = renderResponse.container.getElementsByClassName("cmp-accordion__item");
    expect(panel.length).toBe(1)
    userEvent.click(AddButton[0])
    expect(RemoveButton.length).toBe(2);
    userEvent.click(RemoveButton[1]);
    expect(panel.length).toBe(1)
    expect(RemoveButton.length).toBe(0);
  })
