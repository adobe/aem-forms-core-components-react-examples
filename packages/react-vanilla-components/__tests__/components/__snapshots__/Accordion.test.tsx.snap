// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
<div>
  <div
    class="cmp-accordion "
    data-cmp-enabled="true"
    data-cmp-is="adaptiveFormAccordion"
    data-cmp-single-expansion="true"
    data-cmp-visible="true"
    data-placeholder-text="false"
  >
    <div
      class="cmp-accordion__label-container"
    >
      <label
        class="cmp-accordion__label"
        for="accordion-widget"
      >
        Accordion
      </label>
    </div>
    <div
      aria-live="polite"
    />
    <div
      class="cmp-accordion__item"
      data-cmp-hook-accordion="item"
      id="panelcontainer-541adbc656-widget"
    >
      <h3
        class="cmp-accordion__header"
      >
        <button
          aria-controls="panelcontainer-541adbc656"
          aria-expanded="true"
          class="cmp-accordion__button cmp-accordion__button--expanded"
          data-cmp-hook-accordion="button"
          id="panelcontainer-541adbc656-button"
          role="button"
          type="button"
        >
          <span
            class="cmp-accordion__title"
          >
            Items
          </span>
          <span
            class="cmp-accordion__icon"
          />
        </button>
      </h3>
      <div
        aria-hidden="false"
        aria-labelledby="panelcontainer-541adbc656-button"
        class="cmp-accordion__panel cmp-accordion__panel--expanded"
        data-cmp-hook-adaptiveformaccordion="panel"
        id="panelcontainer-541adbc656-panel"
        role="region"
      />
    </div>
    <div
      class="cmp-accordion__item"
      data-cmp-hook-accordion="item"
      id="panelcontainer-e48ec263e1-widget"
    >
      <h3
        class="cmp-accordion__header"
      >
        <button
          aria-controls="panelcontainer-e48ec263e1"
          aria-expanded="false"
          class="cmp-accordion__button "
          data-cmp-hook-accordion="button"
          id="panelcontainer-e48ec263e1-button"
          role="button"
          type="button"
        >
          <span
            class="cmp-accordion__title"
          >
            Number
          </span>
          <span
            class="cmp-accordion__icon"
          />
        </button>
      </h3>
      <div
        aria-hidden="true"
        aria-labelledby="panelcontainer-e48ec263e1-button"
        class="cmp-accordion__panel cmp-accordion__panel--hidden"
        data-cmp-hook-adaptiveformaccordion="panel"
        id="panelcontainer-e48ec263e1-panel"
        role="region"
      />
    </div>
  </div>
</div>
`;
